src/org/mozilla/javascript/optimizer/DynamicRuntime.java

```diff
@@ -0,0 +1,99 @@
+package org.mozilla.javascript.optimizer;
+
+import java.lang.invoke.CallSite;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import org.mozilla.classfile.ByteCode;
+import org.mozilla.classfile.ClassFileWriter.MHandle;
+import org.mozilla.javascript.ConsString;
+import org.mozilla.javascript.Context;
+import org.mozilla.javascript.Kit;
+import org.mozilla.javascript.ScriptRuntime;
+
+public class DynamicRuntime {
+    public static final String ADD_OP = "add";
+
+    public static final String BOOTSTRAP_SIGNATURE =
+            "(Ljava/lang/invoke/MethodHandles$Lookup;"
+                    + "Ljava/lang/String;"
+                    + "Ljava/lang/invoke/MethodType;"
+                    + ")Ljava/lang/invoke/CallSite;";
+    public static final String METHOD_SIGNATURE_OOC =
+            "(Ljava/lang/Object;Ljava/lang/Object;Lorg/mozilla/javascript/Context;)Ljava/lang/Object;";
+
+    public static final MHandle BOOTSTRAP_HANDLE =
+            new MHandle(
+                    ByteCode.MH_INVOKESTATIC,
+                    "org.mozilla.javascript.optimizer.DynamicRuntime",
+                    "bootstrap",
+                    DynamicRuntime.BOOTSTRAP_SIGNATURE);
+
+    /** Perform an add assuming both arguments are Integer objects and fail otherwise. */
+    public static Object addIntegers(Object o1, Object o2, Context cx) {
```

Comment:
How often do you use `java.lang.Integer` in Rhino?
As far as I know, the result of numeric operations is almost always a `double` or `java.lang.Double`.
https://github.com/mozilla/rhino/blob/3a29b988e7e918ac6478d72c1e6e14af9e25e547/src/org/mozilla/javascript/optimizer/BodyCodegen.java#L3365
Do you mean you want to use `java.lang.Integer` as much as possible in the future?

Reply:
I'm also curious what the motivation was for a special case for Integer addition and for removing ScriptRuntime.wrapNumber from the result of adding two Numbers